clio-dataaccess-model/src/main/scala/org/broadinstitute/clio/server/dataaccess/elasticsearch/ElasticsearchIndex.scala

```diff
@@ -48,6 +56,33 @@ class ElasticsearchIndex[CI <: ClioIndex](
     s"$rootDir$dir/"
   }
 
+  /**
+    * The ID of a json record in elasticsearch. By default, each record's ID consists
+    * of the concatenated values of the Key fields.
+    * Additional fields can be added to the ID using the ElasticsearchIndex constructor.
+    */
+  def getId(json: Json): String = {
+
+    def getAsString(key: String): String = {
+      json.asObject
+        .flatMap(_.apply(key))
+        .flatMap {
+          case s if s.isString =>
+            s.asString
+          case a if a.isArray || a.isObject =>
+            throw new RuntimeException("Arrays and objects cannot be used as ID fields")
+          case j => Option(j.toString())
+        }
+        .getOrElse(throw new RuntimeException(s"Could not get $key from json"))
+    }
+
+    val keyFields = FieldMapper[clioIndex.KeyType].fields.keys
+      .map(snakeCaseTransformation)
+    (keyFields ++ defaults.asObject.map(_.keys).getOrElse(Iterable.empty)).toSeq.sorted
```

Comment:
The more I think about it, the more convinced I am we shouldn't do `sorted`. It'll break the existing ubam index. Last time we rebuilt it it took the better half of a day, and we've backfilled significantly more records since then. We shouldn't take Clio out for a full day unless we really need to.